PostsAndUsersVIPC/Others/Network.swift

```diff
@@ -16,30 +16,35 @@ enum APIError: LocalizedError {
 
 class Network {
     
-    let session: URLSession
-    let decoder: JSONDecoder
+    private let session: URLSession
+    private let decoder: JSONDecoder
     
     init(session: URLSession = URLSession.shared, decoder: JSONDecoder = JSONDecoder()) {
         self.session = session
         self.decoder = decoder
     }
     
-    func fetchData<T: Decodable>(endpoint: APIEndpoint, completion: @escaping (Result<T, APIError>) -> Void) {
+    func fetchData<T: Decodable>(endpoint: APIEndpoint, completion: @escaping (Result<T, APIError>) -> Void) -> URLSessionDataTask? {
         guard let request = buildRequest(endpoint: endpoint) else {
             completion(.failure(.request))
-            return
+            return nil
         }
         let task = session.dataTask(with: request) { data, _, error in
             if let error = error {
                 completion(.failure(.server(error)))
                 return
             }
-            
+
             guard let data = data else {
                 completion(.failure(.empty))
                 return
             }
             
+            if let data = data as? T {
```

Comment:
Se o tipo retornado de fato for um _Data_, você estará fazendo um cast inútil, já que o _Generic_ é substituído em tempo de compilação. O que você poderia fazer é `if data is Data`